### PR TITLE
Supports version and stats command.

### PR DESCRIPTION
**Pull request by bluele**
Hi.
I hope this module supports `version` and `stats` command.

### CODE REVIEW

**Comment by eatstoomuchjam** [comment]
@bradfitz Any plans to merge this soon?  I was just looking for this exact stuff and almost started my own branch to add it.

**Comment by hgfischer** [comment]
:arrow_up: :arrow_up: :arrow_up: :arrow_up:

**Comment by SuperQ** [comment]
@bradfitz Ping :-) It'd be nice to get stats support merged.

**Comment by tomdionysus** [comment]
@bradfitz I second that from @SuperQ, I'd like to use Version as a health check.

**Comment by bluele** [comment]
I signed it!

File: memcache/memcache.go [hunk 1]
```diff
@@ -297,6 +297,33 @@ func (c *Client) FlushAll() error {
...
+	versions := make(map[net.Addr]string)
+	ch := make(chan error, buffered)
+	sn := 0
+	c.selector.Each(func(addr net.Addr) error {
+		sn += 1
+		go func() {
+			ch <- c.versionFromAddr(addr, func(result string) {
+				mu.Lock()
+				defer mu.Unlock()
+				versions[addr] = result
```
Comment (bradfitz) [inline memcache/memcache.go]: this is a data race. You're mutating the versions map from multiple goroutines.

File: memcache/memcache.go [hunk 2]
```diff
@@ -297,6 +297,33 @@ func (c *Client) FlushAll() error {
 	return c.selector.Each(c.flushAllFromAddr)
 }
 
+// Version returns the memcached version from all servers.
+func (c *Client) Version() (map[net.Addr]string, error) {
```
Comment (bradfitz) [inline memcache/memcache.go]: Do people want this signature? I wonder if there should be a plain string return value as well for the case where they all match?

Or maybe it should only return (version string, err error) and it returns an error if any don't match, explaining in the error which have which versions?

File: memcache/memcache.go [hunk 3]
```diff
@@ -319,6 +346,33 @@ func (c *Client) Touch(key string, seconds int32) (err error) {
...
+	stats := make(map[net.Addr]map[string]string)
+	ch := make(chan error, buffered)
+	sn := 0
+	c.selector.Each(func(addr net.Addr) error {
+		sn += 1
+		go func() {
+			ch <- c.statsFromAddr(addr, func(stat map[string]string) {
+				mu.Lock()
+				defer mu.Unlock()
+				stats[addr] = stat
```
Comment (bradfitz) [inline memcache/memcache.go]: another race

File: memcache/memcache.go [hunk 4]
```diff
@@ -319,6 +346,33 @@ func (c *Client) Touch(key string, seconds int32) (err error) {
 	})
 }
 
+// Stats returns the stats from all servers.
+func (c *Client) Stats() (map[net.Addr]map[string]string, error) {
```
Comment (bradfitz) [inline memcache/memcache.go]: We want a better stats type with field names and actual integers, rather than map[string]string.

**Pull request closed without merging**